src/main/java/org/truffleruby/core/proc/ProcNodes.java

```diff
@@ -214,23 +216,45 @@ public DynamicObject binding(DynamicObject proc) {
         @Child private CallBlockNode callBlockNode = CallBlockNode.create();
 
         @Specialization
-        public Object call(DynamicObject proc, Object[] args, NotProvided block) {
-            return callBlockNode.executeCallBlock(
-                    Layouts.PROC.getDeclarationContext(proc),
-                    proc,
-                    ProcOperations.getSelf(proc),
-                    Layouts.PROC.getBlock(proc),
-                    args);
+        public Object call(VirtualFrame frame, DynamicObject proc, Object[] args, NotProvided block) {
+            MaterializedFrame callerFrame = RubyArguments.getCallerFrame(frame);
+            MaterializedFrame declarationFrame = Layouts.PROC.getDeclarationFrame(proc);
+
+            try {
+                return callBlockNode.executeCallBlock(
+                        Layouts.PROC.getDeclarationContext(proc),
+                        proc,
+                        ProcOperations.getSelf(proc),
+                        Layouts.PROC.getBlock(proc),
+                        args);
+            } catch (ReturnException e) {
+                if (declarationFrame.equals(callerFrame)) {
```

Comment:
Based on the Ruby specs it looks like when the proc is called it has to figure out if it's still in the same frame as when it was defined, but I'm really struggling to find a way to make that determination. (That is to say, this `.equals` is definitely wrong. 😂)

```rb
  # spec/ruby/core/proc/new_spec.rb

  it "raises a LocalJumpError when context of the block no longer exists" do
    def some_method
      Proc.new { return }
    end
    res = some_method()

    lambda { res.call }.should raise_error(LocalJumpError)
  end

  it "returns from within enclosing method when 'return' is used in the block" do
    # we essentially verify that the created instance behaves like proc,
    # not like lambda.
    def some_method
      Proc.new { return :proc_return_value }.call
      :method_return_value
    end
    some_method.should == :proc_return_value
  end
```

Reply:
Hello, and sorry for the late reply, I was quite busy recently.

We have a way to track whether a frame is "on stack", and I think we can use that to fix this spec.
In this case, I guess we should check if the *method frame* is on stack when executing the `return`, because that's where it would return.
This is done with the class `FrameOnStackMarker`, and the check should probably be done only in `ReturnNode`, i.e., the Proc executes until the `return` and the `return` raises `LocalJumpError` because it cannot return there.

This is not exactly an easy starting issue, but I can help you go through it if you want to fix it.